app/controllers/users_controller.rb

```diff
@@ -1,31 +1,79 @@
 class UsersController < ApplicationController
-  def show
-    @user = User.find_by id: params[:id]
-    return if @user
-
-    flash[:danger] = t "controllers.users.user_not_found"
-    redirect_to root_path
-  end
+  before_action :logged_in_user, except: [:index, :edit, :update, :destroy]
+  before_action :correct_user, only: %i(edit update)
+  before_action :admin_user, only: :destroy
+  before_action :find_user_by_id, only: :edit
 
-  def new
-    @user = User.new
+  def index
+    @users = User.page(params[:page]).per 20
   end
 
   def create
     @user = User.new user_params
     if @user.save
       log_in @user
       flash[:success] = t "controllers.users.welcome_text"
-      redirect_to @user
+      redirect_back_or user
     else
       flash[:danger] = t "controllers.users.err_text"
       render :new
     end
   end
 
+  def new
+    @user = User.new
+  end
+
+  def edit; end
+
+  def show
+    @user = User.find_by id: params[:id]
+    return if @user
+
+    flash[:danger] = t "controllers.users.user_not_found"
+    redirect_to root_path
+  end
+
+  def update
+    @user = User.find params[:id]
+    if @user.update user_params
+      flash[:success] = t "controllers.users.update_success"
+      redirect_to @user
+    else
+      render "edit"
+    end
+  end
+
+  def destroy
+    User.find(params[:id]).destroy
+    flash[:success] = t "controllers.users.flash_delete"
+    redirect_to users_url
+  end
+
+  def logged_in_user
+    return if logged_in?
+
+    store_location
+    flash[:danger] = t "controllers.users.err_login"
+    redirect_to login_url
+  end
+
+  def admin_user
+    redirect_to root_url unless current_user.admin?
+  end
+
+  def find_user_by_id
+    @user = User.find params[:id]
```

Comment:
vẫn câu hỏi cũ user k tìm thấy thì sao

Reply:
Chị cho em hỏi chút hàm find_user này dùng cho edit update destroy thì luôn luôn có id user trên param rồi mà vẫn phải làm trường hợp không tìm thấy ạ

Reply:
Em có thể sửa id ở trên url phải không nếu em sửa thành 1 id không được phép xem hoặc id đó không tồn tại thì thế nào ???

Reply:
À vâng a, em nghĩ là làm cái Requiring the right user là dùng cho cái find luôn